src/lua/LuaFileSystem.cpp

```diff
@@ -7,15 +7,90 @@
 #include "LuaObject.h"
 #include "Pi.h"
 
-/*
- * Interface: FileSystem
- *
- * A global table that provides access to the filesystem.
- *
- * This interface is protected. It can only be used by scripts in Pioneer's
- * data directory. Mods and scripts in the user directory that try to use it
- * will get a Lua error.
- */
+#include "core/StringUtils.h"
+
+struct SplitURI
+{
+	std::string_view full_path;
+	std::string_view scheme;
+	std::string_view rel_path;
+};
+
+SplitURI ParseURIArg(lua_State* l, int numArg)
+{
+	SplitURI res;
+	res.full_path = LuaPull<std::string_view>(l, numArg);
+	size_t sep_pos = res.full_path.find("://");
+	if (sep_pos != std::string_view::npos)
+	{
+		res.scheme = res.full_path.substr(0, sep_pos);
+		sep_pos += 3;
+		res.rel_path = res.full_path.substr(sep_pos, res.full_path.size() - sep_pos);
+	}
+	return res;
+}
+
+struct FileSource
+{
+	FileSystem::FileSource* source;
+	bool					is_read_write;
+};
+
+FileSource DetermineFileSource(std::string_view scheme)
+{
+	FileSource res;
+	if (compare_ci(scheme, "user"))
+	{
+		res.source = &FileSystem::userFiles;
+		res.is_read_write = true;
+	} else if (compare_ci(scheme, "data"))
+	{
+		res.source = &FileSystem::gameDataFiles;
+		res.is_read_write = false;
+	} else
+	{
+		res.source = nullptr;
+		res.is_read_write = false;
+	}
+	return res;
+}
+
+static lua_CFunction l_original_io_open = nullptr;
+
+void LuaFileSystem::register_raw_io_open_function(lua_CFunction open)
+{
+	l_original_io_open = open;
+}
+
+int LuaFileSystem::l_patched_io_open(lua_State* L)
+{
+	SplitURI split_uri = ParseURIArg(L, 1);
+	const FileSource fs = DetermineFileSource(split_uri.scheme);
+	if (nullptr == fs.source)
+	{
+		return luaL_error(L, "Path '%s' does not contain a valid scheme, must be user:// or data://", split_uri.full_path.data());
+	}
+
+	std::string_view mode_str = LuaPull<std::string_view>(L, 2, "r");
+	if (mode_str.find_first_of("aw+") != std::string_view::npos && !fs.is_read_write) {
+		return luaL_error(L, "File '%s' is read-only.", split_uri.full_path.data());
+	}
+
+	try
+	{
+		::std::string abs_path = fs.source->Lookup(std::string(split_uri.rel_path)).GetAbsolutePath();
+		LuaPush(L, abs_path);
+		lua_replace(L, 1);
+
+		const int rv = l_original_io_open(L);
```

Comment:
I've most likely mentioned this elsewhere, but since I can't remember - this will fail if the file is provided by a `FileSourceZip`, and the failure may not throw an error (will open an incorrect version of the file) if the FileSourceZip is a mod overwriting a game file.

This will have to be addressed before we add lua code which opens any files from the game data directory. Given the purpose of this PR is to provide flexible access to the user directory however, I won't consider it a requirement for merge.

Reply:
Yeah, I'll have a look at that over time, seems like a fair little project and right down my alley.